### DISCUSSION TITLE
Разработать LDAP модуль, для синхронизации сотрудников и автоматического их создания

### DISCUSSION THREAD

**Discussion by jorikfon** [mikopbx/Core#146] [Ideas (Идеи)]
Можно за основу взять презентацию манго
[модуль лдап для mikopbx.pdf](https://github.com/mikopbx/Core/files/6855716/mikopbx.pdf)

**Comment by EDautoAdmin** [comment]
Добрый день.
Прочёл. Было бы неплохо так сделать, однако добавлю пару деталей:
1. Disable в AD - удаление из списка сотрудников.
2. В зависимости от OU выход на провайдера/внешний номер

PS: Жаль, что астер не умеет работать с БД напрямую, типа SQL или LDAP, а работает с текстовых файлом  :( :( :(

**Reply by vfeshenko** [reply]
Вообще-то Астер умеет часть конфигурации хранить в БД, зависит от того, как его настроить. Пользователей точно умеет, сам такую конфигурацию поднимал. Тут больше вопрос, а как это все реализовано в Аскозии/Мико.
Пользователи PBX хранятся в sqlite /cf/conf/mikopbx.db, а вот как дальше они передаются Астериску не копал, подозреваю, что Мико просто генерит текстовые конфиги.

**Reply by jorikfon** [reply]
Да именно так и происходит, MikoPBX генерирует текстовые конфиги. Но проблема не в них, как быть если сотрудник которого удаляем из AD назначен дежурным в блоке нерабочего времени или в маршруте по умолчанию, его просто так выключить не получится не ломая логики звонков.

**Reply by EDautoAdmin** [reply]
Вариантов несколько:
1. Сообщать об ошибке конфигурации вверху, где колокольчик. В полях форм, где присутствовал удаленный сотрудник, писать "ОШИБКА !!!"
2. В конфиге сделать поле "Аварийный телефон", на который будут поступать звонки, куда можно внести любой телефон (Хоть сотовый директора. :) :) :)
3. Не удалять телефонный номер при удалении сотрудника, который участвует в каких-либо действиях. Сообщать админу при синхронизации как ошибку.

Можно ещё что-нибудь придумать. ;)

**Comment by jorikfon** [comment]
Ну вот наверное не удалять такой номер самое логичное, типа если сотрудник не последний агент в очереди, и не используется в тех местах где невозможно просто его удалить то удалить можно без проблем, возможно это будут единичные случаи, когда какого-то основного, дежурного удаляем.

Или сделать синхронизацию по кнопке, админ нажимает и видит ошибки при создании, удалении сотрудников.

**Reply by EDautoAdmin** [reply]
Тогда, перед синхронизацией, нужно сначала провести все возможные проверки, а потом позволять синхронизацию, но, даже в таком случае, возможна ситуация, когда админ телефонии синхронизует, а админ AD удаляет юзера, но это весьма маловероятная ситуация и решается организационно, о чём, к стати, можно выдать предупреждение.

...и к стати, почему в качестве БД юзеров нельзя использовать LDAP AD? Понимаю, что тут подход ещё сложнее в плане отслеживания удаления, но было бы круто. :) Мы, например, не удаляем учётки сотрудников, а отключаем их. Ну и номер может "уходить" к другому сотруднику. LDAP AD тоже интересный вариант.

**Reply by jorikfon** [reply]
Потому что LDAP нужен 5% пользователей MikoPBX, это очень частный случай, я не видел чтобы независимая система так работала, всегда делают синхронизацию.

**Reply by EDautoAdmin** [reply]
На вскидку пруфы. Это из того, что мы используем.

openfire server - http://iantonov.me/page/ustanovka-openfire-jabber-server-v-debian-7-avtorizacija-cherez-active-directory-ldap?ysclid=liy079lvx5546049217
openvpn server - https://openvpn.net/vpn-server-resources/openvpn-access-server-on-active-directory-via-ldap/

PS: Насчёт 5% пользователей, возможно, но это самые платёжеспособные пользователи. ;)
